### PR TITLE
Add experimental option to run unary-stream RPCs on a single Python thread.

### PR DESCRIPTION
**Pull request by gnossen**
Benchmarking revealed that context switching and lock contention led to a considerable amount of overhead in executing RPCs taking the "integrated" code path, which used multiple threads.

Overall, this PR has been shown to consistently decrease end-to-end latency of unary-stream RPCs by around 6.5% on Intel x86-64 workstations. Additional testing involving pinned cores and disabling Intel Turbo revealed that, under non-ideal thermal conditions, this win may degrade to around 2.5%.

An experimental channel option consumed only by the Python layer and not passed to core has been added to activate the single-threaded code path, since it cannot fulfill the `grpc.Future` interface.

This change was tested both in OSS using bazel and in the Google codebase with Blaze, using various sanitizers, including ASAN and MSAN. I was originally concerned that the change to `channel.pyx.pxi` would result in a memory leak, but this was not born out by testing.

At the moment, testing this code path is manual, since activating it across the entire test suite would require manual instrumentation and parametrization of every existing unit test. I have filed [an issue](https://github.com/grpc/grpc/issues/20752) to track this work.

This PR resolves https://github.com/grpc/grpc/issues/20741.

### CODE REVIEW

File: src/python/grpcio/grpc/_channel.py [hunk 1]
```diff
@@ -636,6 +719,55 @@ def future(self,
...
+            credentials=None,
+            wait_for_ready=None,
+            compression=None):
+        deadline = _deadline(timeout)
+        serialized_request = _common.serialize(request,
+                                               self._request_serializer)
+        if serialized_request is None:
+            state = _RPCState((), (), (), grpc.StatusCode.INTERNAL,
+                              'Exception serializing request!')
+            raise _Rendezvous(state, None, None, deadline)
```
[the remark this answers is not in the document]
Comment (gnossen) [reply src/python/grpcio/grpc/_channel.py]: This raises the same exception as the unary-unary code path does. If anything, we should refactor https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_channel.py#L486 to have a more sensible interface.

File: src/python/grpcio/grpc/_channel.py [hunk 2]
```diff
@@ -1042,6 +1174,18 @@ def _augment_options(base_options, compression):
...
+def _separate_channel_options(options):
+    """Separates core channel options from Python channel options."""
+    core_options = []
+    python_options = []
+    for pair in options:
+        if pair[0] == grpc.ChannelOptions.SingleThreadedUnaryStream:
+            python_options.append(pair)
+        else:
+            core_options.append(pair)
+    return python_options, core_options
```
[the remark this answers is not in the document]
Comment (gnossen) [reply src/python/grpcio/grpc/_channel.py]: I'm not sure how this comment relates to this line.

File: src/python/grpcio_tests/tests/stress/unary_stream_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    ('grpc.max_metadata_size', 16 * 1024 * 1024),
+    ('grpc.max_receive_message_length', 64 * 1024 * 1024),
+    (grpc.ChannelOptions.SingleThreadedUnaryStream, 1),
+]
+
+
+@contextlib.contextmanager
+def _running_server():
+    server_process = subprocess.Popen(
+        [sys.executable, '-c', _SERVER_CODE],
```
[the remark this answers is not in the document]
Comment (gnossen) [reply src/python/grpcio_tests/tests/stress/unary_stream_benchmark.py]: This is not intended to be run in Google3. Note that it was not added to anything that will get picked up by Blaze or by `setup.py`, for that matter. For the moment, this is intended to be a manual test run under Bazel.

File: src/python/grpcio/grpc/_channel.py [hunk 4]
```diff
@@ -267,7 +298,152 @@ def cancel(self):
...
+                return True
+
+    def initial_metadata(self):
+        """See grpc.Call.initial_metadata"""
+        with self._state.condition:
+
+            def _done():
+                return self._state.initial_metadata is not None
+
+            _common.wait(self._state.condition.wait, _done)
```
[the remark this answers is not in the document]
Comment (gnossen) [reply src/python/grpcio/grpc/_channel.py]: I don't agree with this. If we tried to move the RPC forward with these methods, we would be fundamentally mulithreaded. The use case here is for other threads spawned by the application to be able to block on these attributes being made available by the single donated thread running through `SingleThreadedRendezvous`.

File: src/python/grpcio/grpc/__init__.py [hunk 5]
```diff
@@ -1953,6 +1953,17 @@ class Compression(enum.IntEnum):
...
 
+class ChannelOptions(object):
+    """Indicates a channel option unique to gRPC Python.
+
+     This enumeration is part of an EXPERIMENTAL API.
+
+     Attributes:
+       SingleThreadedUnaryStream: Perform unary-stream RPCs on a single thread.
+  """
+    SingleThreadedUnaryStream = "SingleThreadedUnaryStream"
```
[the remark this answers is not in the document]
Comment (lidizheng) [reply src/python/grpcio/grpc/__init__.py]: As discussed offline, moving it to the experimental folder to enable us to change it in the future.

File: src/python/grpcio/grpc/_channel.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lidizheng) [reply src/python/grpcio/grpc/_channel.py]: Let's link the issue here

**Comment by gnossen** [comment]
The TODO was added just a few lines down.

File: src/python/grpcio_tests/tests/unit/_metadata_test.py [hunk 6]
```diff
@@ -202,6 +202,9 @@ def testUnaryUnary(self):
     def testUnaryStream(self):
         multi_callable = self._channel.unary_stream(_UNARY_STREAM)
         call = multi_callable(_REQUEST, metadata=_INVOCATION_METADATA)
+        # NOTE(gnossen): In the single-threaded case, we must consume at least
+        # one message before the initial metadata will show up.
+        next(call)
```
[the remark this answers is not in the document]
Comment (gnossen) [reply src/python/grpcio_tests/tests/unit/_metadata_test.py]: https://github.com/grpc/grpc/issues/20762

File: src/python/grpcio/grpc/_channel.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gnossen) [reply src/python/grpcio/grpc/_channel.py]: https://github.com/grpc/grpc/issues/20763

**Comment by gnossen** [comment]
Known failures:
 - https://github.com/grpc/grpc/issues/20436
 - https://github.com/grpc/grpc/issues/18111

**Pull request merged**